### PR TITLE
Feature/nested context

### LINKED ISSUE #938: Nested Contexts to allow easier hierarchical tests
To make it easier to structure the tests (as described in this video by Kevlin Henney: https://www.youtube.com/watch?v=tWn8RA_DEic) it would be great to allow nesting of contexts:

```sql
create or replace package queue_spec as
  -- %suite(Queue Spec)
  
  -- %context(a_not_empty_queue)
    
    -- %context(that_is_not_full)
      -- %test
      procedure becomes_longer_when_non_null_value_enqueued;
      -- %test
      procedure becomes_full_when_enqueued_up_to_capacity;
    -- %endcontext
    
    -- %context(that_is_full)
      -- %test
      procedure ignores_further_enqueued_values;
      -- %test
      procedure becomes_non_full_when_dequeued;
    -- %endcontext

    -- %test
    procedure dequeues_values_in_order_enqueued;

  -- %endcontext
end;
```
Leading to
```
Queue Spec
  a_not_empty_queue
    that_is_not_full
      becomes_longer_when_non_null_value_enqueued [SUCCESS]
      becomes_full_when_enqueued_up_to_capacity [SUCCESS]
    that_is_full
      ignores_further_enqueued_values [SUCCESS]
      becomes_non_full_when_dequeued [SUCCESS]
    dequeues_values_in_order_enqueued [SUCCESS]
```

**What to do in case of missing endcontext?**
Write a warning

### LINKED ISSUE #966: Use of fullstop character in  --%context(<name>)
### Describe the bug
Using a full stop character in ```--%context(<name>) ```  will cause all other contexts in the suite to be executed under an entirely different suite as seen in the output section below. The reason for this I believe is because the way a context name is interpreted as a path to allow users to execute specific contexts. 

### Version Info
**utPLSQL:**  v3.1.6.2735
**SQL Developer:**  19.1.0.094

### To Reproduce
**1. Create tests**
```sql
create or replace package test_context_bug_1 is

  --%suite

  --%context(suite_1 context_1)

    --%test
    procedure test1;

  --%endcontext
  
  --%context(suite_1 context_2)

    --%test
    procedure test2;

  --%endcontext

end;
/

create or replace package body test_context_bug_1 is

  procedure test1 is
  begin
    ut.expect( 1 ).to_equal( 1 );
  end;

  procedure test2 is
  begin
    ut.expect( 1 ).to_equal( 1 );
  end;

end;
/

create or replace package test_context_bug_2 is

  --%suite
  
  --%context(suite_2 context_1)
  
    --%test
    procedure test1;

  --%endcontext
  
  --%context(suite_2 context_2)

    --%test
    procedure test2;

  --%endcontext

end;
/

create or replace package body test_context_bug_2 is

  procedure test1 is
  begin
    ut.expect( 1 ).to_equal( 1 );
  end;

  procedure test2 is
  begin
    ut.expect( 1 ).to_equal( 1 );
  end;

end;
/
 ```
**2. Add a full stop in a context name**
``` sql 
--%context(suite_2. context_2)  
```

**3. Run tests**
``` sql
SET SERVEROUTPUT ON SIZE UNLIMITED
EXECUTE ut.run(); 
```

### Outputs
**Expected output:**
```
test_context_bug_1
  suite_1 context_2
    test2 [.003 sec]
  suite_1 context_1
    test1 [.002 sec]
 
test_context_bug_2
  suite_2 context_2
    test2 [.002 sec]
  suite_2 context_1
    test1 [.002 sec]
 
Finished in .013272 seconds
4 tests, 0 failed, 0 errored, 0 disabled, 0 warning(s)
PL/SQL procedure successfully completed.
```

**Actual output:**
```
test_context_bug_1
  suite_2 context_1
    test1 [.003 sec]
  suite_1 context_2
    test2 [.001 sec]
  suite_1 context_1
    test1 [.001 sec]
 
test_context_bug_2
  suite_2. context_2
    test2 [.003 sec]
 
Finished in .01466 seconds
4 tests, 0 failed, 0 errored, 0 disabled, 0 warning(s)
PL/SQL procedure successfully completed.
```

### Expected behavior
When a full stop is present in ```--%context(<name>) ```  the other contexts in the suite shouldn't be executed under another suite.

### LINKED ISSUE #1016: Consistent syntax for suite/test/context annotations
**Current state**

Currently `--%suite` and `--%test` annotations take `description` as argument.
The `--%context` annotation doesn't allow that and expects `name` to be given instead.

This is inconsistent and leads to confusion when using the framework.

Engineers expect simplicity and ease of use from the framework which isn't the case with `--%context` annotation.

Consider the below example:
```sql
create or replace package queue_spec as
  --%suite(Queue specification)

  --%context(A new queue)

    --%test(Cannot be created with non positive bounding capacity)
    procedure non_positive_bounding_cap;

  --%endcontext
end;
```

With this package, current implementation of utPLSQL gives the following:

| suitepath | description | name |
|-----------|------------|------|
| queue_spec | Queue specification | queue_spec |
| queue_spec.A new queue | A new queue | A new queue |
| queue_spec.A new queue.non_positive_bounding_cap | Cannot be created with non positive bounding capacity | non_positive_bounding_cap |

- The name of suite comes from package name and description from annotation
- The name of test comes from procedure name and description from annotation 
- The name of context is take from annotation and description is copied from name

The above package spec is not really valid, as context name should not contain any whitespace or full stop characters. Context name rules should conform to naming rules for Oracle objects.

utPLSQL does not give any warning on invalid context names too.

The proper implementation of the above package following current framework behavior should be more like below.
```sql
create or replace package queue_spec as
  --%suite(Queue specification)

  --%context(a_new_queue)
  --%displayname(A new queue)

    --%test(Cannot be created with non positive bounding capacity)
    procedure non_positive_bounding_cap;

  --%endcontext
end;
```

The above is confusing, hard to remember and easy to confuse engineers.

| suitepath | description | name |
|-----------|------------|------|
| queue_spec | Queue specification | queue_spec |
| queue_spec.a_new_queue | A new queue | a_new_queue |
| queue_spec.a_new_queue.non_positive_bounding_cap | Cannot be created with non positive bounding capacity | non_positive_bounding_cap |

**Describe the solution you'd like**

I suggest to change the current behavior so that the `--%context` annotation accepts description as argument and is aligned with `--%test` and `--suite` annotations.

Framework would need to provide a way for giving an explicit name to a context as well as allow for implicit naming, if users don't care about context names. 

Rules:
- if no `name` is specified for a context, the context name is defaulted to `nested_context_#N` where `N` is the consecutive number of the context inside it's parent.
- if a `name` is specified for the context, the name is used instead
- if provided `name` is not valid, a warning is given and the name is defaulted to `nested_context_#N` 

Given the original example.
```sql
create or replace package queue_spec as
  --%suite(Queue specification)

  --%context(A new queue)

    --%test(Cannot be created with non positive bounding capacity)
    procedure non_positive_bounding_cap;

  --%endcontext
end;
```

The new implementation would give the following results:

| suitepath | description | name |
|-----------|------------|------|
| queue_spec | Queue specification | queue_spec |
| queue_spec.nested_context_#1 | A new queue | nested_context_#1 |
| queue_spec.nested_context_#1.non_positive_bounding_cap | Cannot be created with non positive bounding capacity | non_positive_bounding_cap |

To give an explicit name to the context, users would use syntax:
```sql
create or replace package queue_spec as
  --%suite(Queue specification)

  --%context(A new queue)
  --%name(a_new_queue)

    --%test(Cannot be created with non positive bounding capacity)
    procedure non_positive_bounding_cap;

  --%endcontext
end;
```

**Describe alternatives you've considered**

An alternative would be to convert the description into context name by:
- making it lower-case
- trimming leading and trailing whitespace and full stop characters
- replacing all the whitespaces, full stops and special characters with underscores

This could work pretty well for some many use-cases.

```sql
create or replace package queue_spec as
  --%suite(Queue specification)

  --%context(A new queue)

    --%test(Cannot be created with non positive bounding capacity)
    procedure non_positive_bounding_cap;

  --%endcontext
end;
```

The alternative implementation would give the following results:

| suitepath | description | name |
|-----------|------------|------|
| queue_spec | Queue specification | queue_spec |
| queue_spec.a_new_queue | A new queue | a_new_queue |
| queue_spec.a_new_queue.non_positive_bounding_cap | Cannot be created with non positive bounding capacity | non_positive_bounding_cap |

We would still need to use default context names if no description is provided.

### PR DESCRIPTION
**Pull request by jgebal**
Added ability to nest contexts in suites - Resolves #938 
Added fix for invalid context name - Resolves #966

Changed approach for context naming - Resolves #1016 

So now you can write tests like a Pro (see [talk](https://www.youtube.com/watch?v=tWn8RA_DEic) by Kevlin Henney)
```sql
create or replace package queue_spec as
  -- %suite(Queue Spec)
  
  -- %context(a_not_empty_queue)
    
    -- %context(that_is_not_full)
      -- %test
      procedure becomes_longer_when_non_null_value_enqueued;
      -- %test
      procedure becomes_full_when_enqueued_up_to_capacity;
    -- %endcontext
    
    -- %context(that_is_full)
      -- %test
      procedure ignores_further_enqueued_values;
      -- %test
      procedure becomes_non_full_when_dequeued;
    -- %endcontext

    -- %test
    procedure dequeues_values_in_order_enqueued;

  -- %endcontext
end;
```
Leading to
```
Queue Spec
  a_not_empty_queue
    that_is_not_full
      becomes_longer_when_non_null_value_enqueued [SUCCESS]
      becomes_full_when_enqueued_up_to_capacity [SUCCESS]
    that_is_full
      ignores_further_enqueued_values [SUCCESS]
      becomes_non_full_when_dequeued [SUCCESS]
    dequeues_values_in_order_enqueued [SUCCESS]
```

### CODE REVIEW

File: source/core/ut_suite_builder.pkb [hunk 1]
```diff
@@ -798,27 +807,17 @@ create or replace package body ut_suite_builder is
     end loop;
   end;
 
-  procedure warning_on_incomplete_context(
+  procedure warning_on_floating_endcontext(
```
Comment (jgebal) [inline source/core/ut_suite_builder.pkb]: warning_on_extra_endcontext

**Comment by jgebal** [comment]
@pesse, @lwasylow, @Pazus  - I'd like to get your feedback on my doubts around:
`--%context(name)` vs `--%context(description)` on SLACK before merging/altering this PR.

**Comment by jgebal** [comment]
@pesse, @lwasylow - the PR is ready for review.
I would definitely appreciate review of documentation changes.

File: source/core/ut_suite_builder.pkb (hunk 1 above)
Comment (jgebal) [reply source/core/ut_suite_builder.pkb]: done

**Comment by lwasylow** [comment]
I see that it's implemented but I think that nested context is misuse. I think it should be replaced by nested suites and context should be not nested and exists within a suite. I will look into code 👍

**Comment by jgebal** [comment]
I was of the same opinion from the moment we originally discussed contexts.
But users want it and there are situations where it is actually useful.

**Pull request merged**